transactions/WFLY-17742_SupportTransactionsDuringGracefulShutdown

```diff
@@ -0,0 +1,155 @@
+* Integrating with the existing logging mechanism of WildFly, users should be notified when subsystems are delaying WildFly’s suspension. This is especially true when a negative timeout is used (i.e. indefinite graceful shutdown). Moreover, internal details like 'ServerActivity' do not need to be prominent. Part of working out such a notification would be how to consistently describe a ServerActivity to a user in an understandable manner without exposing internal technical implementation details.
+  * ~~Notifications from the Transactions subsystem should explicitly report enough information to identify the transactions that are delaying the graceful shutdown of WildFly. Moreover, notifications should report whether the transaction is in-doubt or heuristic.~~ This is already implemented in the Transactions subsystem, i.e. details about transactions that need to be recovered (automatically or manually) are reported using the existing logging mechanisms of WildFly.
+
+=== Nice-to-Have Requirements
+
+* Introducing an asynchronous graceful shutdown would enable all ServerActivity implementations to receive a pre-suspend/suspend signal concurrently. In this way, the timeout will have the same duration for all ServerActivity implementations.
+* [The sequence to suspend](https://github.com/wildfly/wildfly-core/blob/main/server/src/main/java/org/jboss/as/server/suspend/SuspendController.java#L57-L99) ServerActivity implementations during WildFly’s graceful shutdown should be Last Input First Output (LIFO), i.e. the last ServerActivity implementation that was registered during startup should be the first one to get suspended.
+
+=== Non Requirements
+
+* As discussed previously, in a cloud environment, when the hosting entity crashes, its state should be recovered by the cloud environment without the intervention of the user
+
+== Implementation Plan
+
+*Aim of the proposal*. All subsystems that have the ability to start transactions should modify their suspension handlers to properly delay WildFly’s graceful shutdown as long as there are transactions to complete. Of course, users should be notified when subsystems are delaying WildFly’s suspension. This is especially true when a negative timeout is used (i.e. indefinite graceful shutdown). Moreover, internal details like 'ServerActivity' do not need to be prominent. Part of working out such a notification would be how to consistently describe a ServerActivity to a user in an understandable manner without exposing internal technical implementation details.
+
+=== Narayana
+
+Narayana needs to expose more APIs to give third-party more information about transactions. Moreover, Narayana’s Recovery Coordinator should be developed to offer more support to complete in-doubt transactions during its suspension.
```

Comment:
Here I think that in-doubt is enough. Heuristic transactions need the intervention of the admin to get resolved.

Reply:
We do need to log the existence of heuristics before shutting down, with sufficient information to enable a developer to manually resolve them. Elsewhere you mention requiring that enough information is reported to identify the transactions that are delaying the graceful shutdown. You didn't define what a notification is though, would include an entry in the log or something more?

Reply:
line 68 `Integrating with the existing logging mechanism of WildFly, users should be notified when subsystems are delaying WildFly’s suspension`. Simplifying, notifications are logs in WildFly's output.

Reply:
I understand, but when a heuristic transaction is present it should also  prevent the subsystem being shut down. So the question is - is Narayana API ready for that or does it also require changes like "in-doubt txns scenario"?

Reply:
Yep, a heuristic outcome should be considered as a blocker during the graceful shutdown. I added this information in the proposal following your comments. To reply to your question, I don't think Narayana needs a new API to handle heuristic txns during the graceful shutdown. The presence of heuristic transactions will be notified to the user/admin as the reason why the graceful shutdown hasn't completed yet. Then the user/admin should take care of those heuristic transactions manually if they want to complete the graceful shutdown properly.

Reply:
A heuristic outcome must not block shutdown since it is an unrecoverable situation.
Just ensure that sufficient information gets logged to enable the admin to manually recover it or ignore it as not a problem.

Reply:
Here, "logged" can include  keeping the transaction logs around.

Reply:
> A heuristic outcome must not block shutdown since it is an unrecoverable situation. Just ensure that sufficient information gets logged to enable the admin to manually recover it or ignore it as not a problem.

If a heuristic outcome won't block shutdown, we are back to square one.

NOTE: The reasons why this proposal was presented can be found in paragraph 1.1 of this proposal.

Reply:
@jmfinelli - what it seems like is being considered is that so long as the logs are somehow made available to the admin (maybe not the raw Narayana object store log) then the node could be recycled. Like offload a summary of the heuristic to the admin to deal with outside of graceful shutdown.

So I think it's rather not that it doesn't block shutdown, but rather that it only blocks shutdown long enough to make the information the admin would need available post node shutdown.

E.g. `tx 1 for node A recorded at [some agreed timestamp, maybe the cluster?] resulted in a heuristic hazard where resource X committed and resource Y rolledback. The branch ID for resource X was [XidA] and the branch ID for resource Y was [XidB]`

I am not sure where logging for these heuristics might be offloaded to though, perhaps there is some ACID durable store of this kind of data available in OpenShift somewhere or we could mandate a database be provided perhaps?

Reply:
> @jmfinelli - what it seems like is being considered is that so long as the logs are somehow made available to the admin (maybe not the raw Narayana object store log) then the node could be recycled. Like offload a summary of the heuristic to the admin to deal with outside of graceful shutdown.
> 
> So I think it's rather not that it doesn't block shutdown, but rather that it only blocks shutdown long enough to make the information the admin would need available post node shutdown.
> 
> E.g. `tx 1 for node A recorded at [some agreed timestamp, maybe the cluster?] resulted in a heuristic hazard where resource X committed and resource Y rolledback. The branch ID for resource X was [XidA] and the branch ID for resource Y was [XidB]`

Thanks for the clarification.

> I am not sure where logging for these heuristics might be offloaded to though, perhaps there is some ACID durable store of this kind of data available in OpenShift somewhere or we could mandate a database be provided perhaps?

In case we go for a solution available in OpenShift, we end up with an ad-hoc solution that can be used only in OpenShift (similar to the K8s Operator). As we already know, this kind of ad-hoc solutions are not easily portable. About using a database, I will think about that and I'll add a reply ASAP. I would just mention that waiting indefinitely for the resolution of heuristic transactions is something that already happens in the K8s Operator. If you think there are good reasons to change this behaviour, I think that we need to start investigating this and, potentially, change the K8s operator ASAP (better if in a dedicated JIRA/proposal). IMO, given the fact that this behaviour has already been validated through multiple rounds of development and has been extensively tested (across two versions of the K8s Operator) I think it is safer not to change this behaviour here.

Reply:
I don't think it is necessary to change what is happening already, so that is a good consideration to add to the requirements or non-requirements (that shutdown will not be optimized further than existing function for heuristics). That said, it can be a useful improvement to consider.

Without knowing this for sure, but for illustrative purposes the decision to not block shutdown would be to free up the pod earlier. But we would still need to keep the logs as long (as the administrator wants them). OTOH something to consider is it may (somehow) prevent the admin from deterministically handling (as well as they can) the heuristic if a pod name starts to get re-used before the heuristics are tidied up....

Reply:
> waiting indefinitely for the resolution of heuristic transactions

A heuristic outcome, by definition, cannot be resolved, unless you mean a human steps in and fixes up the system manually, ie it cannot happen automatically. Perhaps you are referring to in doubt transactions.

Reply:
> > waiting indefinitely for the resolution of heuristic transactions
> 
> A heuristic outcome, by definition, cannot be resolved, unless you mean a human steps in and fixes up the system manually, ie it cannot happen automatically. Perhaps you are referring to in doubt transactions.

Yes, I mean that a human being needs to step in and fix up the system manually.

Reply:
If we just offload just some logging rather than the object store, it would mean any existing tooling that uses the object store to control the resource manager to forget heuristics would not be able to be used.

Reply:
I am not sure of the better word that resolve, but indeed there may be no manual resolution or reconciliation possible. What is happening is we are notifying that we are waiting for an administrator to ask us to forget the heuristic branches of resource managers that are in conflict with the decision of the transaction manager.

Reply:
> Without knowing this for sure, but for illustrative purposes the decision to not block shutdown would be to free up the pod earlier. But we would still need to keep the logs as long (as the administrator wants them). OTOH something to consider is it may (somehow) prevent the admin from deterministically handling (as well as they can) the heuristic if a pod name starts to get re-used before the heuristics are tidied up....

While I can see the value in freeing up a pod earlier, I don't believe it justifies the design and development effort or the cost of maintaining external resources* given that heuristic outcomes are quite rare in production environments. On the other hand, delaying the shutdown of WildFly in cases of heuristic branches seems like a calculated risk, based on the observed frequency of such outcome. Moreover, in this way, administrators are provided with all necessary tools and information available within WildFly to easily address heuristic branches.

* The cost of storing logs in cloud resources should be considered very carefully. For instance, maintaining a database just for the purpose of storing infrequent events can lead to substantial expenses, hard to justify. Moreover, various cloud providers may offer ad-hoc methods for storing events/logs, which makes me wonder whether we should consider different solutions for different cloud providers.

Reply:
Is it possible to provide the option to automatically offload the ObjectStore to storage or block if not configured?

For Kubernetes, perhaps that would be a PVC (assuming it's ACID storage)? And if not configured to do so then it would just block as the default.

Reply:
> * The cost of storing logs in cloud resources should be considered very carefully. For instance, maintaining a database just for the purpose of storing infrequent events can lead to substantial expenses, hard to justify. Moreover, various cloud providers may offer ad-hoc methods for storing events/logs, which makes me wonder whether we should consider different solutions for different cloud providers.

These kinds of statement are worrying, heuristics can break the C word (ACID) and the consequences of doing that can be significant. Just ensure that you've logged the information somewhere - either in a log file that the customer or the platform backs up (something which all providers support) or elsewhere. But don't block shutdown waiting for a human being to step in to resolve the heuristic, nobody is going to buy that solution - hence the requirement to maintain a record of the heuristic but not to require that it gets resolved by the platform. In terms of proposal, just state the requirements not the solutions.

Reply:
What is the relationship between this proposal and https://github.com/wildfly/wildfly-proposals/pull/295 - does it replace that one and can we close it?

Reply:
@mmusgrov I was back here and going to mention something about that 295 too, thanks for raising it :)

Reply:
@mmusgrov I was back here and going to mention something about that 295 too, thanks for raising it :)

I think that when considering what to do about Heuristics we should consider the existing documentation and tooling for this for Narayana on bare metal: https://www.narayana.io//docs/product/index.html#manage_transaction

If we delete the ObjectStore, the tooling that we have can't be used to send the xa_forget to the RM. I am not saying we have to keep it just to do that but the tooling is something our users should be aware of already and an alternative approach would need to be shared.

Reply:
In terms of the requirement (which could lead into solution, but let's keep to requirement wording):

Do people think it is important that the transaction manager node ID is not re-used until the heuristics has been forgotten by the resource manager? If not, why not? I am not sure I have an answer but it seems like something that we need to know.

Does the user need to know anything except the branch ID of the heuristically complete branch? I presume they needs to know details about the resource manager so they can find who had the branch, but do they also need to know about the other branches that were in the transaction? I am not sure, but maybe it could help to know (e.g. branch TX1:branch1 in the message broker committed and TX1:branch2 in the database rolled back?) currently I expect our tooling would show this. But I am not sure if you can take that TX1:branch1 to the message broker and find out what it did in that transaction in order to try to reconcile. If you can though it seems definitely interesting to retain that too.

Reply:
> If we delete the ObjectStore, the tooling that we have can't be used to send the xa_forget to the RM. I am not saying we have to keep it just to do that but the tooling is something our users should be aware of already and an alternative approach would need to be shared

Maybe we can investigate serialising the transaction record and  logging that using the "WildFly logging mechanism". The the end user can use [our tooling to forget heuristics](https://github.com/jbosstm/narayana/blob/main/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/tools/log/LogBrowser.java).

Reply:
> Do people think it is important that the transaction manager node ID is not re-used until the heuristics has been forgotten by the resource manager? If not, why not? I am not sure I have an answer but it seems like something that we need to know.

If it's reused then the heuristic branch becomes eligible for rollback (via the NodeNameXAResourceOrphanFilter). That action may well compound the problem that caused the original heuristic outcome. So, yes I think it is important not to reuse the node ID.

Reply:
> Does the user need to know anything except the branch ID of the heuristically complete branch? I presume they needs to know details about the resource manager so they can find who had the branch, but do they also need to know about the other branches that were in the transaction? I am not sure, but maybe it could help to know (e.g. branch TX1:branch1 in the message broker committed and TX1:branch2 in the database rolled back?) currently I expect our tooling would show this. But I am not sure if you can take that TX1:branch1 to the message broker and find out what it did in that transaction in order to try to reconcile. If you can though it seems definitely interesting to retain that too.

The customer needs to be able to work out the consequences of the heuristic and so develop an action plan to resolve the situation.
That means we need the other branches involved in the heuristic and we need to know what we told them to do; for example if transaction committed and a branch reported heuristic commit then the customer needs to consider that the branch might have exposed its state before the commit phase (the I in ACID) and likewise if the branch reported heuristic rollback then he needs to consider the consequences of that too (the C in ACID). Etc for the full
matrix of outcomes (XA_HEUxxx).

So basically, we need the full transaction record - see my earlier comment about an option for storing the full record using the "WildFly logging mechanism".

Reply:
> > Do people think it is important that the transaction manager node ID is not re-used until the heuristics has been forgotten by the resource manager? If not, why not? I am not sure I have an answer but it seems like something that we need to know.
> 
> If it's reused then the heuristic branch becomes eligible for rollback (via the NodeNameXAResourceOrphanFilter). That action may well compound the problem that caused the original heuristic outcome. So, yes I think it is important not to reuse the node ID.

OK, but I think for OpenShift to prevent the node ID being reused it will be easiest to block the pod as that should be a 1:1 mapping. Though that would be implementation choice (for that environment) and thus something for later (after the requirements themselves are agreed).

Reply:
Also consider:

1. JTS support since we maintain separate records for participants (type name "/CosTransactions/XAResourceRecord").
2. I believe WFTC also logs something in order to support EJB remoting requirements.

These could be handled as not supported. But not supporting remote EJB calls could be an issue so we should ask the WFTC team to review this so that they know to delay shutdown, according to the requirements of this proposal, if there are pending transactions.